provider/lpseal/task_sdr.go

```diff
@@ -0,0 +1,220 @@
+			c, err := cid.Parse(p.PieceCID)
+			if err != nil {
+				return false, xerrors.Errorf("parsing piece cid: %w", err)
+			}
+
+			pieceInfos[i] = abi.PieceInfo{
+				Size:     abi.PaddedPieceSize(p.PieceSize),
+				PieceCID: c,
+			}
+		}
+
+		commd, err = nonffi.GenerateUnsealedCID(sectorParams.RegSealProof, pieceInfos)
+		if err != nil {
+			return false, xerrors.Errorf("computing CommD: %w", err)
+		}
+	} else {
+		commd = zerocomm.ZeroPieceCommitment(abi.PaddedPieceSize(ssize).Unpadded())
+	}
+
+	sref := storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(sectorParams.SpID),
+			Number: abi.SectorNumber(sectorParams.SectorNumber),
+		},
+		ProofType: sectorParams.RegSealProof,
+	}
+
+	// get ticket
+	maddr, err := address.NewIDAddress(uint64(sectorParams.SpID))
+	if err != nil {
+		return false, xerrors.Errorf("getting miner address: %w", err)
+	}
+
+	// FAIL: api may be down
+	// FAIL-RESP: rely on harmony retry
+	ticket, ticketEpoch, err := s.getTicket(ctx, maddr)
+	if err != nil {
+		return false, xerrors.Errorf("getting ticket: %w", err)
+	}
+
+	// do the SDR!!
+
+	// FAIL: storage may not have enough space
+	// FAIL-RESP: rely on harmony retry
+
+	// LATEFAIL: compute error in sdr
+	// LATEFAIL-RESP: Check in Trees task should catch this; Will retry computing
+	//                Trees; After one retry, it should return the sector to the
+	// 			      SDR stage; max number of retries should be configurable
+
+	err = s.sc.GenerateSDR(ctx, sref, ticket, commd)
+	if err != nil {
+		return false, xerrors.Errorf("generating sdr: %w", err)
+	}
+
+	// store success!
+	n, err := s.db.Exec(ctx, `UPDATE sectors_sdr_pipeline
+		SET after_sdr = true, ticket_epoch = $3, ticket_value = $4
+		WHERE sp_id = $1 AND sector_number = $2`,
+		sectorParams.SpID, sectorParams.SectorNumber, ticketEpoch, []byte(ticket))
+	if err != nil {
+		return false, xerrors.Errorf("store sdr success: updating pipeline: %w", err)
+	}
+	if n != 1 {
+		return false, xerrors.Errorf("store sdr success: updated %d rows", n)
+	}
+
+	return true, nil
+}
+
+func (s *SDRTask) getTicket(ctx context.Context, maddr address.Address) (abi.SealRandomness, abi.ChainEpoch, error) {
+	ts, err := s.api.ChainHead(ctx)
+	if err != nil {
+		return nil, 0, xerrors.Errorf("getting chain head: %w", err)
+	}
+
+	ticketEpoch := ts.Height() - policy.SealRandomnessLookback
+	buf := new(bytes.Buffer)
+	if err := maddr.MarshalCBOR(buf); err != nil {
+		return nil, 0, xerrors.Errorf("marshaling miner address: %w", err)
+	}
+
+	rand, err := s.api.StateGetRandomnessFromTickets(ctx, crypto.DomainSeparationTag_SealRandomness, ticketEpoch, buf.Bytes(), ts.Key())
+	if err != nil {
+		return nil, 0, xerrors.Errorf("getting randomness from tickets: %w", err)
+	}
+
+	return abi.SealRandomness(rand), ticketEpoch, nil
+}
+
+func (s *SDRTask) CanAccept(ids []harmonytask.TaskID, engine *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	// todo check storage (reserve too?)
```

Comment:
Oh yeah, this is somewhat important.

Buut it will need a slight change to CanAccept - basically:
* SDRTask.CanAccept would check if it can accept the task (so, enough space is available in storage)
* It would then reserve the space (this mechanism exists in the storage subsystem already, so that's easy)
* But then, if harmony doesn't actually claim the task, it would need to call some callback from CanAccept to drop the reservation

The last point can be as simple as an additional return from CanAccept, like `claimed func(didClaim bool)error`